File: src/main/java/org/starlisp/core/LispStream.java
package org.starlisp.core;

import java.io.IOException;

public interface LispStream {
  void writeJavaString(String str);
  void writeJavaChar(char ch) throws IOException;
  boolean eof();
  boolean close() throws IOException;
  LispObject read() throws IOException;
  LispChar readChar() throws IOException;
}

File: src/main/scala/org/starlisp/core/LispStream.scala
package org.starlisp.core

import java.io.{PrintWriter, StringWriter, IOException}

trait LispStream extends LispObject {
  def eof: Boolean
  @throws(classOf[IOException]) def close: Boolean
}

trait LispInputStream extends LispStream {
  @throws(classOf[IOException]) def read: LispObject
  @throws(classOf[IOException]) def readChar: LispChar
}

trait LispOutputStream extends LispStream {
  @throws(classOf[IOException]) def write(str: String)
  @throws(classOf[IOException]) def write(ch: Char)
}

class StringOutputStream(writer: StringWriter = new StringWriter) extends LispOutputStreamImpl(new PrintWriter(writer)) {
  def getOutputStreamString: String = {
    val sb = writer.getBuffer
    val result: String = sb.toString
    sb.setLength(0)
    result
  }
}

